### PR TITLE
Add support for param mutation under inference mode

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Summary:
In HF model rwkv, we have parameter mutation under inference mode which should be safe. This PR does multiple things to make sure it works:
1. We execute global autograd mutation while tracing so that we can actually trace through parameter inplace mutation 
2. Add support for parameter mutation under inference mode in AOTAutograd 
3. Add support for parameter mutation under inference mode in export.

Test Plan:
test

Rollback Plan:

Differential Revision: D79460136

cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

### CODE REVIEW

File: torch/_functorch/_aot_autograd/input_output_analysis.py [hunk 1]
```diff
@@ -460,6 +460,7 @@ def create_graph_signature(
         named_buffers=buffer_names,
         num_user_inputs=num_user_args,
         num_user_outputs=num_user_fw_outs,
+        trace_joint=trace_joint,
```
Comment (tugsbayasgalan) [inline torch/_functorch/_aot_autograd/input_output_analysis.py:463]: cc: @ezyang does this look ok in terms of your refactoring work?

File: torch/fx/experimental/proxy_tensor.py [hunk 2]
```diff
@@ -1430,9 +1430,11 @@ def __torch_function__(
                 torch.amp.autocast_mode._exit_autocast,
             ]:
                 node.meta["val"] = None
+            # For autocast, the python APIs run so we don't have to run them again 
+            # here. 
+            if func is torch._C._set_grad_enabled:
```
Comment (tugsbayasgalan) [inline torch/fx/experimental/proxy_tensor.py:1435]: cc: @zou3519, @bdhirsh This is behavioral change from before.

**Review by zou3519 — COMMENTED** [review]
your tests don't look happy

File: torch/fx/experimental/proxy_tensor.py (hunk 2 above)
Comment (zou3519) [reply torch/fx/experimental/proxy_tensor.py:1435]: er what is going on here?

Comment (tugsbayasgalan) [reply torch/fx/experimental/proxy_tensor.py:1435]: We are actually running the state change instead of just proxy-ing so that param mutation under no-grad still works.

File: test/functorch/test_aotdispatch.py [hunk 3]
```diff
@@ -5368,7 +5368,7 @@ def forward(self, x):
             RuntimeError,
             "Found a graph input that requires gradients, and received a mutation",
         ):
-            aot_export_module(mod, [inp], trace_joint=False)
+            aot_export_module(mod, [inp], trace_joint=True)
```
Comment (ydwu4) [inline test/functorch/test_aotdispatch.py]: Why we need the change here?

File: test/export/test_export.py [hunk 4]
```diff
@@ -325,6 +325,36 @@ def forward(self, *args):
...
+    %div : [num_users=2] = call_function[target=torch.ops.aten.div.Tensor](args = (%p_parameter, 2), kwargs = {})
+    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%x, %div), kwargs = {})
+    return (div, add)""",
+        )
+
+        self.assertTrue("div" in val.keys())
+        self.assertTrue("parameter" in val.values())
+
+        with self.assertRaisesRegex(RuntimeError, "leaf"):
+            ep.module()(torch.rand(4, 4))
```
Comment (ydwu4) [inline test/export/test_export.py]: Is the error that leaf receives a mutation? Does it work if we set the parameters.require_grad to be False?

File: torch/fx/experimental/proxy_tensor.py (hunk 2 above)
Comment (zou3519) [reply torch/fx/experimental/proxy_tensor.py:1435]:
```
# Don't actually run the function! We just want to trace the calls
# into a graph. We don't actually want to change global autograd state.
```

Comment (zou3519) [reply torch/fx/experimental/proxy_tensor.py:1435]: So, at the very least, pre-dispatch tracing should not change the state after it runs. If there is only one single torch.set_grad_enabled(False) call in the graph, after one does the pre-dispatch trace, it should not be off.

File: torch/export/graph_signature.py [hunk 5]
```diff
@@ -121,6 +121,7 @@ class OutputKind(Enum):
     USER_OUTPUT = auto()
     LOSS_OUTPUT = auto()
     BUFFER_MUTATION = auto()
+    PARAMETER_MUTATION = auto()
```
Comment (ydwu4) [inline torch/export/graph_signature.py:124]: Does this gets properly serialized and de-serialized? Curious why this doesn't trigger the error [here](https://github.com/pytorch/pytorch/blob/main/torch/_export/serde/serialize.py#L1279).  We need to add the serialize and deserialize logic as well.

BC-wise, this could be made BC-preserving i.e. a minor version upgrade. The idea is to add a new  optional field parameter_mutation in schema.OutputSpec.  Deserializer should be able to automatically assign it None when de-serializing old artifact.

File: test/export/test_export.py (hunk 4 above)
Comment (tugsbayasgalan) [reply test/export/test_export.py]: Yeah we can do that but we don't know when it is supposed to be True or False which depends on whether we are under torch.no_grad or something else.

File: torch/fx/experimental/proxy_tensor.py (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/fx/experimental/proxy_tensor.py:1435]: Yep i wrapped it in ctx manager under export.

File: test/functorch/test_aotdispatch.py [hunk 6]
```diff
@@ -5364,11 +5364,15 @@ def forward(self, x):
 
         mod = M()
         inp = torch.randn(2, requires_grad=True)
-        with self.assertRaisesRegex(
```
Comment (ydwu4) [inline test/functorch/test_aotdispatch.py:5367]: It seems there's a divergence between aot_export_module vs aot_export_joint_simple for mutating an input that requires gradient, aot_export_module can export but not aot_export_joint_simple? Is this intentional?

File: torch/_export/serde/serialize.py [hunk 7]
```diff
@@ -1241,6 +1242,15 @@ def serialize_output_spec(self, spec: ep.OutputSpec) -> OutputSpec:
                     buffer_name=spec.target,
                 )
             )
+        elif spec.kind == ep.OutputKind.PARAMETER_MUTATION:
```
Comment (ydwu4) [inline torch/_export/serde/serialize.py:1245]: I didn't see the logic of handling de-serializing a parameter mutation OutputSpec. Why this doesn't trigger an error for the serde test?

File: torch/_export/serde/schema.py [hunk 8]
```diff
@@ -359,6 +365,7 @@ class OutputSpec(_Union):
     gradient_to_user_input: Annotated[GradientToUserInputSpec, 50]
     user_input_mutation: Annotated[UserInputMutationSpec, 60]
     token: Annotated[OutputTokenSpec, 70]
+    parameter_mutation: Annotated[Optional[ParameterMutationSpec], 80]
```
Comment (ydwu4) [inline torch/_export/serde/schema.py]: After a second thought, I think a union type is probably already optional lol meaning that when we can always create the dataclass (using the new dataclass definition) from json of an old artifact. Should we remove the Optional? 

This is BC-compitable but it breaks FC. I remember seeing internal pipelines sometimes rely on FC not sure if it's fixed or not. So probably we should check with internal folks to warn them or do more testing also cc @yiming0416 @SherlockNoMad

File: test/export/test_export.py (hunk 4 above)
Comment (ydwu4) [reply test/export/test_export.py]: the requires_grad attribute of tensors doesn't change with with torch.no_grad i think. One way to test is to manually flip the attribute to false and see if eager works. 

It seems weird that eager doesn't work but we can export successfully lol, which seems a divergent behavior and is hard to explain.

File: test/functorch/test_aotdispatch.py (hunk 6 above)
Comment (tugsbayasgalan) [reply test/functorch/test_aotdispatch.py:5367]: Yeah i think no one uses aot_export_joint_simple and for some reason this API takes joint option as boolean lol

File: torch/_functorch/_aot_autograd/input_output_analysis.py (hunk 1 above)
Comment (ezyang) [reply torch/_functorch/_aot_autograd/input_output_analysis.py:463]: The change looks reversible enough, but I can't really say I understand why you need yet another flag here. Additionally in your tests you are not even tracing out the joint!

Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/input_output_analysis.py:463]: This is because in the graph signature, we need to distinguish between joint IR vs inference IR. In the joint IR case, we can't have mutations on parameters.

File: torch/export/_trace.py [hunk 9]
```diff
@@ -1674,7 +1674,24 @@ def override_getattribute_for_subclasses(args):
...
+                old_state = torch.is_grad_enabled()
+                try:
+                    yield
+                finally:
+                    torch._C._set_grad_enabled(old_state)
+
+            with (
+                ctx,
+                override_getattribute_for_subclasses(flat_args),
+                _maybe_restore_grad_state(),
```
Comment (ydwu4) [inline torch/export/_trace.py:1693]: Is it accurate that parameter mutation is supported in export (for all the flags/mode)? We should document all the semantic of mutations, e.g. input, buffer, parameters mutations somewhere. Right now I don't have a clear mental model for them.

**Review by ydwu4 — APPROVED** [review]

File: torch/_functorch/_aot_autograd/input_output_analysis.py (hunk 1 above)
Comment (ezyang) [reply torch/_functorch/_aot_autograd/input_output_analysis.py:463]: One thing I'll say is that for a generic "this works when torch.compile works" export frontend, these cases also need to be accounted for.
